rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/scan/operation/RecurseDirectoryAndAddFiles.java

```diff
@@ -56,8 +63,13 @@ private void recurseAndAddFiles(FileService fileService, FileModel file)
             {
                 for (File reference : subFiles)
                 {
+                    if (FileDiscoveredListenerUtil.shouldSkip(listeners, new DefaultFileDiscoveredEvent(reference.getAbsolutePath())))
+                        continue;
+
                     FileModel subFile = fileService.createByFilePath(file, reference.getAbsolutePath());
-                    recurseAndAddFiles(fileService, subFile);
+                    for (FileDiscoveredListener listener : listeners)
+                        listener.fileModelCreated(event, context, subFile);
```

Comment:
That would be another approach. The biggest downside is that it would have to iterate them again, which may be redundant.